src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,386 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/usage_manager.h"
+
+#include "config/configuration.h"
+#include "kafka/server/logger.h"
+#include "storage/api.h"
+#include "vlog.h"
+
+namespace kafka {
+
+static constexpr std::string_view period_key{"period"};
+static constexpr std::string_view max_duration_key{"max_duration"};
+static constexpr std::string_view buckets_key{"buckets"};
+
+static bytes key_to_bytes(std::string_view sv) {
+    bytes k;
+    k.append(reinterpret_cast<const uint8_t*>(sv.begin()), sv.size());
+    return k;
+}
+
+struct persisted_state {
+    std::chrono::seconds configured_period;
+    size_t configured_windows;
+    std::vector<usage_window> current_state;
+};
+
+static ss::future<>
+persist_to_disk(storage::kvstore& kvstore, persisted_state s) {
+    using kv_ks = storage::kvstore::key_space;
+
+    co_await kvstore.put(
+      kv_ks::usage,
+      key_to_bytes(period_key),
+      serde::to_iobuf(s.configured_period));
+    co_await kvstore.put(
+      kv_ks::usage,
+      key_to_bytes(max_duration_key),
+      serde::to_iobuf(s.configured_windows));
+    co_await kvstore.put(
+      kv_ks::usage,
+      key_to_bytes(buckets_key),
+      serde::to_iobuf(s.current_state));
+}
+
+static std::optional<persisted_state>
+restore_from_disk(storage::kvstore& kvstore) {
+    using kv_ks = storage::kvstore::key_space;
+    std::optional<iobuf> period, windows, data;
+    try {
+        period = kvstore.get(kv_ks::usage, key_to_bytes(period_key));
+        windows = kvstore.get(kv_ks::usage, key_to_bytes(max_duration_key));
+        data = kvstore.get(kv_ks::usage, key_to_bytes(buckets_key));
+    } catch (const std::exception& ex) {
+        vlog(
+          klog.debug,
+          "Encountered exception when retriving usage data from disk: {}",
+          ex);
+        return std::nullopt;
+    }
+    if (!period && !windows && !data) {
+        /// Data didn't exist
+        return std::nullopt;
+    } else if (!period || !windows || !data) {
+        vlog(
+          klog.warn,
+          "Inconsistent usage_manager on disk state detected, failed to "
+          "recover state");
+        return std::nullopt;
+    }
+    return persisted_state{
+      .configured_period = serde::from_iobuf<std::chrono::seconds>(
+        std::move(*period)),
+      .configured_windows = serde::from_iobuf<size_t>(std::move(*windows)),
+      .current_state = serde::from_iobuf<std::vector<usage_window>>(
+        std::move(*data))};
+}
+
+static ss::future<> clear_persisted_state(storage::kvstore& kvstore) {
+    using kv_ks = storage::kvstore::key_space;
+    try {
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(period_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(max_duration_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(buckets_key));
```

Comment:
I am curious how downstream systems would react to responses that contain intervals that aren't roughly equivelant. I can ask Alena if this is even an issue, if not then this can be removed.

Reply:
we agreed in one of the early meetings that dropping data on config change is acceptable so let's stick with that and we can change it in later releases if that's not a problem. If we ever decide to change e.g. the window size, there will have to be more changes in the stack to support that so for now, this behavior is OK 👍